### PR TITLE
simplity impl to own its onw state

### PR DESCRIPTION
**Pull request by lsem**

### CODE REVIEW

File: bounded_async_foreach.hpp [hunk 1]
```diff
@@ -3,78 +3,78 @@
...
+void bounded_async_foreach(unsigned n, Container c, Callback cb,
+                           FinishCallback finished_cb) {
+  if (n == 0) {
+    finished_cb(make_error_code(std::errc::no_child_process));
+    return;
+  } else if (std::distance(std::cbegin(c), std::cend(c)) == 0) {
+    // nothing has beem processed succesfully.
+    finished_cb(std::error_code());
+    return;
+  }
```
Comment (lsem) [inline bounded_async_foreach.hpp]: commenting multiple lines

Comment (lsem) [reply bounded_async_foreach.hpp]: @oyefremov

File: bounded_async_foreach.hpp [hunk 2]
```diff
@@ -3,78 +3,78 @@
...
-auto bounded_async_foreach(unsigned n, Container c, Callback cb,
-                           FinishedCallback finished) {
-  using iter_t = decltype(std::cbegin(c));
+template <typename Container, typename Callback, typename FinishCallback>
+void bounded_async_foreach(unsigned n, Container c, Callback cb,
+                           FinishCallback finished_cb) {
+  if (n == 0) {
+    finished_cb(make_error_code(std::errc::no_child_process));
+    return;
+  } else if (std::distance(std::cbegin(c), std::cend(c)) == 0) {
```
Comment (oyefremov) [inline bounded_async_foreach.hpp]: is it a fancy way to say
`std::empty(c)`

Comment (lsem) [reply bounded_async_foreach.hpp]: is it supposed to be available for all collections?

Comment (lsem) [reply bounded_async_foreach.hpp]: done.

**Pull request merged**